From f6eda0481b801c78fc5cf269506c161bc8cd396b Mon Sep 17 00:00:00 2001
From: Tony Crisci <tony@…>
Date: Fri, 25 Apr 2014 05:37:03 -0400
Subject: [PATCH] Implement the window::fullscreen event

The fullscreen event is a window with the "change" property set to
"fullscreen". This event should be emitted whenever a window becomes
fullscreen and whenever the window becomes non-fullscreen.

This event can be used to turn off dpms off when a window is fullscreen
or display the fullscreen container name in the status line for instance.
---
 src/con.c                               |  3 ++
 testcases/t/224-ipc-window-fullscreen.t | 59 +++++++++++++++++++++++++++++++++
 2 files changed, 62 insertions(+)
 create mode 100644 testcases/t/224-ipc-window-fullscreen.t

diff --git a/src/con.c b/src/con.c
index 5a3c88d..f1c54b8 100644
--- a/src/con.c
+++ b/src/con.c
@@ -609,6 +609,9 @@ void con_toggle_fullscreen(Con *con, int fullscreen_mode) {
 
     DLOG("mode now: %d\n", con->fullscreen_mode);
 
+    /* Send an ipc window "fullscreen" event */
+    ipc_send_window_event("fullscreen", con);
+
     /* update _NET_WM_STATE if this container has a window */
     /* TODO: when a window is assigned to a container which is already
      * fullscreened, this state needs to be pushed to the client, too */
diff --git a/testcases/t/224-ipc-window-fullscreen.t b/testcases/t/224-ipc-window-fullscreen.t
new file mode 100644
index 0000000..7bb6d69
--- /dev/null
+++ b/testcases/t/224-ipc-window-fullscreen.t
@@ -0,0 +1,59 @@
+#!perl
+# vim:ts=4:sw=4:expandtab
+#
+# Please read the following documents before working on tests:
+# • http://build.i3wm.org/docs/testsuite.html
+#   (or docs/testsuite)
+#
+# • http://build.i3wm.org/docs/lib-i3test.html
+#   (alternatively: perldoc ./testcases/lib/i3test.pm)
+#
+# • http://build.i3wm.org/docs/ipc.html
+#   (or docs/ipc)
+#
+# • http://onyxneon.com/books/modern_perl/modern_perl_a4.pdf
+#   (unless you are already familiar with Perl)
+#
+# Tests that the ipc window::fullscreen event works properly
+#
+# Bug still in: 4.7.2-135-g7deb23c
+use i3test;
+
+my $i3 = i3(get_socket_path());
+$i3->connect()->recv;
+
+my $cv;
+my $t;
+
+sub reset_test {
+    $cv = AE::cv;
+    $t = AE::timer(0.5, 0, sub { $cv->send(0); });
+}
+
+reset_test;
+
+$i3->subscribe({
+        window => sub {
+            my ($e) = @_;
+            if ($e->{change} eq 'fullscreen') {
+                $cv->send($e->{container});
+            }
+        },
+    })->recv;
+
+my $window = open_window;
+
+cmd 'fullscreen';
+my $con = $cv->recv;
+
+ok($con, 'got fullscreen window event (on)');
+is($con->{fullscreen_mode}, 1, 'window is fullscreen');
+
+reset_test;
+cmd 'fullscreen';
+$con = $cv->recv;
+
+ok($con, 'got fullscreen window event (off)');
+is($con->{fullscreen_mode}, 0, 'window is not fullscreen');
+
+done_testing;
-- 
1.9.2

